contracts/product/CachedSetValuer.sol

```diff
@@ -0,0 +1,286 @@
+        uint256 _maxStaleness
+    )
+        public
+    {
+        require(
+            _controller.isResource(address(_setValuer)),
+            "SetValuer must be enabled on Controller"
+        );
+        
+        if (address(_rebasingModule) != address(0)) {
+            require(
+                _controller.isModule(address(_rebasingModule)),
+                "RebasingModule must be enabled on Controller"
+            );
+        }
+        
+        controller = _controller;
+        setValuer = _setValuer;
+        rebasingModule = _rebasingModule;
+        maxStaleness = _maxStaleness;
+    }
+
+    /* ============ External Functions ============ */
+
+    /**
+     * Gets cached valuation if fresh, otherwise calculates new valuation.
+     * Syncs rebasing components before calculation if cache is stale and rebasing module is set.
+     *
+     * @param _setToken        SetToken to get valuation for
+     * @param _quoteAsset      Asset to quote valuation in
+     * @return uint256         SetToken valuation in terms of quote asset
+     */
+    function calculateSetTokenValuation(
+        ISetToken _setToken,
+        address _quoteAsset
+    )
+        external
+        returns (uint256)
+    {
+        require(
+            controller.isResource(address(setValuer)),
+            "SetValuer must be enabled on Controller"
+        );
+
+        CachedValuation storage cachedValuation = cachedValuations[_setToken][_quoteAsset];
+        
+        if (_isCacheValid(cachedValuation.timestamp)) {
+            return cachedValuation.value;
+        }
+
+        // Sync rebasing components before getting new valuation if module is set
+        if (address(rebasingModule) != address(0)) {
+            rebasingModule.sync(_setToken);
+        }
+        
+        uint256 newValuation = setValuer.calculateSetTokenValuation(_setToken, _quoteAsset);
+        
+        _updateCache(_setToken, _quoteAsset, newValuation);
+        
+        return newValuation;
+    }
+
+    /**
+     * Force updates cache with current valuation.
+     * 
+     * @param _setToken        SetToken to update cache for
+     * @param _quoteAsset      Asset to quote valuation in
+     */
+    function updateCache(
+        ISetToken _setToken,
+        address _quoteAsset
+    )
+        external
+    {
+        if (address(rebasingModule) != address(0)) {
+            rebasingModule.sync(_setToken);
+        }
+        
+        uint256 newValuation = setValuer.calculateSetTokenValuation(_setToken, _quoteAsset);
+        
+        _updateCache(_setToken, _quoteAsset, newValuation);
+    }
+
+    /**
+     * Returns both cached and current valuations for comparison.
+     * Does not modify state or update cache.
+     * 
+     * @param _setToken        SetToken to get valuation for
+     * @param _quoteAsset      Asset to quote valuation in
+     * @return ValuationInfo   Struct containing cached and current valuations
+     */
+    function previewValuation(
```

Comment:
Is the `currentValue` returned by the setValuer not 100% accurate without first calling sync on the rebasing module?

Reply:
yes this is correct - added a warning here [`6b24709`](https://github.com/IndexCoop/index-protocol/pull/51/commits/6b24709a0e15ad587d91d05cf835de5fb420d28f)

this function could be updated to simulate the accounting update while still not writing to state